mypyc/ir/rtypes.py

```diff
@@ -436,6 +445,123 @@ def deserialize(cls, data: JsonDict, ctx: 'DeserMaps') -> 'RTuple':
+            items = list(typ.types)
+        elif isinstance(typ, RStruct):
+            items = typ.types
+        elif isinstance(typ, RUnion):
+            items = typ.items
+        else:
+            assert False, "invalid rtype for computing alignment"
+        max_alignment = max([compute_rtype_alignment(item) for item in items])
+        return max_alignment
+
+
+def compute_rtype_size(typ: RType) -> int:
+    """Compute unaligned size of rtype"""
+    platform_size = 4 if IS_32_BIT_PLATFORM else 8
+    if isinstance(typ, RPrimitive):
+        return typ.size
+    elif isinstance(typ, RTuple):
+        return compute_aligned_offsets_and_size(list(typ.types))[1]
+    elif isinstance(typ, RUnion):
+        return compute_aligned_offsets_and_size(typ.items)[1]
+    elif isinstance(typ, RStruct):
+        return compute_aligned_offsets_and_size(typ.types)[1]
+    elif isinstance(typ, RInstance):
+        return platform_size
+    else:
+        assert False, "invalid rtype for computing size"
+
+
+def compute_aligned_offsets_and_size(types: List[RType]) -> Tuple[List[int], int]:
+    """Compute offsets and total size of a list of types after alignment
+
+    Note that the types argument are types of values that are stored
+    sequentially with platform default alignment.
+    """
+    platform_alignment = 4 if IS_32_BIT_PLATFORM else 8
+    unaligned_sizes = [compute_rtype_size(typ) for typ in types]
+    alignment = [compute_rtype_alignment(typ) for typ in types]
+
+    current_offset = 0
+    offsets = []
+    final_size = 0
+    for i in range(len(unaligned_sizes)):
+        offsets.append(current_offset)
+        if i + 1 < len(unaligned_sizes):
+            cur_size = unaligned_sizes[i]
+            next_size = unaligned_sizes[i + 1]
+            if next_size > cur_size:
+                current_offset += cur_size
+                next_alignment = alignment[i + 1]
+                padding = (next_alignment - (current_offset % next_alignment)) % next_alignment
+                current_offset += padding
+            else:
+                current_offset += cur_size
+        else:
+            final_size = current_offset + unaligned_sizes[i]
+            final_padding = ((platform_alignment - (final_size % platform_alignment))
+                             % platform_alignment)
+            final_size += final_padding
+    return offsets, final_size
+
+
+class RStruct(RType):
+    """Represent CPython structs"""
+    def __init__(self,
+                 name: str,
+                 names: List[str],
+                 types: List[RType]) -> None:
+        self.name = name
+        self.names = names
+        self.types = types
+        self.offsets, self.size = compute_aligned_offsets_and_size(types)
+        self._ctype = self.name
+
+    def accept(self, visitor: 'RTypeVisitor[T]') -> T:
+        return visitor.visit_rstruct(self)
+
+    def __str__(self) -> str:
+        # if not tuple(unamed structs)
+        if len(self.names) > 0:
```

Comment:
I think that we can always require names to be defined, for consistency. It's always possible to use some dummy names like `item0`, `item1`, etc. if there are no better names.